blockchain/message_pool/src/msgpool/selection.rs

```diff
@@ -0,0 +1,592 @@
+use super::{allow_negative_chains, create_message_chains, MessagePool, Provider};
+use crate::{run_head_change, Error};
+use address::Address;
+use blocks::Tipset;
+use message::SignedMessage;
+use num_bigint::BigInt;
+use std::cmp::Ordering;
+use std::collections::HashMap;
+
+type Pending = HashMap<Address, HashMap<u64, SignedMessage>>;
+
+impl<T> MessagePool<T>
+where
+    T: Provider + std::marker::Send + std::marker::Sync + 'static,
+{
+    /// Selects messages for including in a block.
+    pub async fn select_messages(
+        &mut self,
+        ts: Tipset,
+        _tq: f64,
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let cur_ts = self.cur_tipset.read().await.clone();
+        // TODO: Implement a more optimal message selection to pack more msgs into a block
+        self.select_messages_greedy(cur_ts.as_ref(), ts).await
+    }
+
+    async fn select_messages_greedy(
+        &mut self,
+        cur_ts: &Tipset,
+        ts: Tipset,
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let base_fee = self.api.read().await.chain_compute_base_fee(&ts)?;
+
+        // 0. Load messages from the target tipset; if it is the same as the current tipset in
+        //    the mpool, then this is just the pending messages
+        let mut pending = self.get_pending_messages(&cur_ts, &ts).await?;
+
+        if pending.is_empty() {
+            return Ok(Vec::new());
+        }
+        // 0b. Select all priority messages that fit in the block
+        // TODO: Implement guess gas
+        let min_gas = 1298450;
+        let (mut result, mut gas_limit) = self
+            .select_priority_messages(&mut pending, &base_fee, &ts)
+            .await?;
+
+        // check if block has been filled
+        if gas_limit < min_gas {
+            return Ok(result);
+        }
+        // 1. Create a list of dependent message chains with maximal gas reward per limit consumed
+        let mut chains = Vec::new();
+        for (actor, mset) in pending.into_iter() {
+            chains.append(
+                &mut create_message_chains(&self.api, &actor, &mset, &base_fee, &ts).await?,
+            );
+        }
+
+        // 2. Sort the chains
+        chains.sort_by(|a, b| b.compare(&a));
+        if !allow_negative_chains(cur_ts.epoch())
+            && !chains.is_empty()
+            && chains[0].curr().gas_perf < 0.0
+        {
+            return Ok(result);
+        }
+        // 3. Merge the head chains to produce the list of messages selected for inclusion, subject to
+        //    the block gas limit.
+        let mut last = chains.len();
+        for i in 0..chains.len() {
+            if !allow_negative_chains(cur_ts.epoch()) && chains[i].curr().gas_perf < 0.0 {
+                break;
+            }
+
+            if chains[i].curr().gas_limit <= gas_limit {
+                gas_limit -= chains[i].curr().gas_limit;
+                result.append(&mut chains[i].curr().msgs.clone());
```

Comment:
This should probably be an `extend` call as well

Reply:
Whats the diff with extend vs append in this context

Reply:
Well more importantly you're allocating a new vector just to drain it into the result, when all you need to do is extend by cloning the elements. I don't know what specifically gets optimized by llvm here, but idk if it can avoid the reallocation of the vector this way

Reply:
`append` is useful when you want to append the elements of another vector that you want to empty and keep around (iirc its capacity is maintained as well), but that's a pretty specific use case where you have a buffer that you keep filling, or something like that. In general `extend` is the method to reach for

Reply:
MMMMMMMM, yeah ok that makes a lot of sense. I will do that :D